### PR TITLE
Feature/disable prefetch fixes

### PR DESCRIPTION
**Pull request by Dkhodos**
*Note*: Before submitting this pull request, please review our [contributing guidelines](https://docs.celeryq.dev/en/main/contributing.html).

## Description

Introduce an option to completely disable broker prefetching without requiring `acks_late`. When enabled, the worker only fetches a new task when an execution slot is available. This prevents head‑of‑line blocking and starvation caused by long‑running tasks.

- Real‑world need: we run long‑lasting tasks (hours) across multiple workers/pods. Prefetch caused busy workers to hold tasks in reserve while idle workers sat empty, leading to severe starvation. Disabling prefetch resolves this by fetching only when a slot is free.
- This work revives and supersedes the stalled discussion and implementation in [PR #9818](https://github.com/celery/celery/pull/9818). All reviewer suggestions from that PR have been applied here.

### What’s included

- New config: `worker_disable_prefetch` (default: False)
- New CLI flag: `--disable-prefetch`
- Core logic: when enabled, wrap `channel.qos.can_consume` to check `reserved_requests` against effective concurrency:
  - use autoscaler `max_concurrency` if present, otherwise pool size
- Documentation updates:
  - user guide (optimizing, configuration) and FAQ explain usage and impact
- Tests:
  - unit tests for consumer/autoscale behavior (including autoscaling edge cases)
  - pytest‑click based CLI test covering `--disable-prefetch`
- Minor nit applied: compute limit via `getattr(c.controller, "max_concurrency", None) or c.pool.num_processes`

### Backward compatibility

- Off by default; existing behavior unchanged.
- When enabled, only reservation timing changes (fetch on free slot); no protocol or API changes.

### Implementation notes

- This is a pragmatic Celery‑side QoS guard. Longer‑term, the ideal place for this behavior would be in Kombu’s QoS, with Celery surfacing the option.

### Testing

- Unit suites for consumer/autoscale/CLI pass locally.
- Docs build and configuration references updated.

### References

- Prior discussion/implementation: [PR #9818](https://github.com/celery/celery/pull/9818)

### CODE REVIEW

**Comment by Dkhodos** [comment]
> ## [Codecov](https://app.codecov.io/gh/celery/celery/pull/9863?dropdown=coverage&src=pr&el=h1&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=celery) Report
> ❌ Patch coverage is `92.85714%` with `1 line` in your changes missing coverage. Please review. ✅ Project coverage is 78.64%. Comparing base ([`33eb148`](https://app.codecov.io/gh/celery/celery/commit/33eb14852310996b1909c8388cd319809d6c8626?dropdown=coverage&el=desc&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=celery)) to head ([`48b254b`](https://app.codecov.io/gh/celery/celery/commit/48b254b47e029fc9bc40fabf831ff980779b9ec6?dropdown=coverage&el=desc&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=celery)).
> 
> [Files with missing lines](https://app.codecov.io/gh/celery/celery/pull/9863?dropdown=coverage&src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=celery)	Patch %	Lines
> [celery/bin/worker.py](https://app.codecov.io/gh/celery/celery/pull/9863?src=pr&el=tree&filepath=celery%2Fbin%2Fworker.py&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=celery#diff-Y2VsZXJ5L2Jpbi93b3JrZXIucHk=)	66.66%	[0 Missing and 1 partial ⚠️ ](https://app.codecov.io/gh/celery/celery/pull/9863?src=pr&el=tree&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=celery)
> Additional details and impacted files
> [☔ View full report in Codecov by Sentry](https://app.codecov.io/gh/celery/celery/pull/9863?dropdown=coverage&src=pr&el=continue&utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=celery). 📢 Have feedback on the report? [Share it here](https://about.codecov.io/codecov-pr-comment-feedback/?utm_medium=referral&utm_source=github&utm_content=comment&utm_campaign=pr+comments&utm_term=celery).

should be covered well now

**Review by Nusnus — COMMENTED** [review]
Please note the integration tests are currently broken in `main`, preventing merges of non-trivial changes.

A fix is in progress 🙏

**Review by avigaffa — APPROVED** [review]

**Review by sashu2310 — APPROVED** [review]

**Comment by bakwc** [comment]
Thanks a lot for working on this! Any chances to get it merged in the nearest future?

**Comment by Dkhodos** [comment]
> Thanks a lot for working on this! Any chances to get it merged in the nearest future?

@Nusnus is CI fixed?

**Comment by Nusnus** [comment]
> > Thanks a lot for working on this! Any chances to get it merged in the nearest future?
> 
> @Nusnus is CI fixed?

Hopefully by (my) EOD today.
I'm going to try a different approach, let's see if it works 🙏

File: docs/userguide/configuration.rst [hunk 1]
```diff
@@ -3131,16 +3131,36 @@ workers, note that the first worker to start will receive four times the
...
+command line flag.
 
 For more on prefetching, read :ref:`optimizing-prefetch-limit`
 
 .. note::
 
     Tasks with ETA/countdown aren't affected by prefetch limits.
 
+.. setting:: worker_disable_prefetch
+
```
Comment (auvipy) [inline docs/userguide/configuration.rst:3202]: versionadd annotation is missing.

**Review by auvipy — CHANGES_REQUESTED** [review]
also, do we have relevant issues open in kombu for the following note:

This is a pragmatic Celery‑side QoS guard. Longer‑term, the ideal place for this behavior would be in Kombu’s QoS, with Celery surfacing the option.

File: docs/userguide/configuration.rst (hunk 1 above)
Comment (Dkhodos) [reply docs/userguide/configuration.rst:3202]: .. versionadded:: 5.4 ?

added.

**Comment by nasis** [comment]
Needed
Looking for it to be merged.

**Review by amirgur — APPROVED** [review]

**Review by nasis — APPROVED** [review]

File: docs/userguide/configuration.rst [hunk 2]
```diff
@@ -3180,16 +3180,38 @@ workers, note that the first worker to start will receive four times the
...
+``worker_disable_prefetch``
+~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+.. versionadded:: 5.4
```
Comment (auvipy) [inline docs/userguide/configuration.rst]:
```suggestion
.. versionadded:: 5.6
```

File: docs/userguide/configuration.rst (hunk 1 above)
Comment (auvipy) [reply docs/userguide/configuration.rst:3202]: 5.6. done

**Review by auvipy — APPROVED** [review]

**Comment by Dkhodos** [comment]
> also, do we have relevant issues open in kombu for the following note:
> 
> This is a pragmatic Celery‑side QoS guard. Longer‑term, the ideal place for this behavior would be in Kombu’s QoS, with Celery surfacing the option.

Added
https://github.com/celery/kombu/issues/2353

Might even have a PR for this later today :) 


Side question:
Wen can we expect a package release with this fix? what number would it be 4.6.?

**Comment by auvipy** [comment]
v5.6

**Comment by auvipy** [comment]
we got a regression report of this. can you please check it? we might need to revert it for a while if needed https://github.com/celery/celery/issues/9913

**Comment by bakwc** [comment]
May be you could leave it at least for redis?

**Comment by auvipy** [comment]
yes, that is possible
